src/ScriptBuilder/Saga/SagaDefinitionReader.cs

```diff
@@ -38,6 +45,192 @@ public static bool TryGetSqlSagaDefinition(TypeDefinition type, out SagaDefiniti
         return true;
     }
 
+    static bool TryInferSagaDefinition(TypeDefinition type, out SagaDefinition definition)
+    {
+        var baseType = type.BaseType as GenericInstanceType;
+        definition = null;
+
+        // Class must directly inherit from NServiceBus.Saga<T> so that no tricks can be pulled from an intermediate class
+        if (baseType == null || !baseType.FullName.StartsWith("NServiceBus.Saga") || baseType.GenericArguments.Count != 1)
+        {
+            return false;
+        }
+
+        CheckIsValidSaga(type);
+
+        string correlationId = null;
+        string transitionalId = null;
+        string tableSuffix = null;
+
+        var attribute = type.GetSingleAttribute("NServiceBus.Persistence.Sql.SqlSagaAttribute");
+        if (attribute != null)
+        {
+            var args = attribute.ConstructorArguments;
+            correlationId = (string)args[0].Value;
+            transitionalId = (string)args[1].Value;
+            tableSuffix = (string)args[2].Value;
+        }
+
+        var sagaDataType = GetSagaDataTypeFromSagaType(type);
+
+        if (correlationId == null)
+        {
+            correlationId = InferCorrelationId(type, sagaDataType);
+        }
+
+        SagaDefinitionValidator.ValidateSagaDefinition(correlationId, type.FullName, transitionalId, tableSuffix);
+
+        if (tableSuffix == null)
+        {
+            tableSuffix = type.Name;
+        }
+
+        definition = new SagaDefinition
+        (
+            correlationProperty: BuildConstraintProperty(sagaDataType, correlationId),
+            transitionalCorrelationProperty: BuildConstraintProperty(sagaDataType, transitionalId),
+            tableSuffix: tableSuffix,
+            name: type.FullName
+        );
+        return true;
+    }
+
+    static string InferCorrelationId(TypeDefinition type, TypeDefinition sagaDataType)
```

Comment:
I read this method a few times and I don't feel comfortable reviewing this unless there is a test for each if statement here. Lots or arcane magic.

@bording has this brilliant idea that we should use a console tool for generating the scripts instead of a msbuild task. This way we could load the actual `dll` into the process and execute the `ConfigureHowToFindSaga` method. 

How about in this iteration we just bring back the support of `Saga` but make the attribute mandatory? The next step would be to drop the msbuild & cecil magic in favour of a simple console tool.

Reply:
I second that opinion, this method is the direct reason why I wanted to get more eyes on this PR.

Reply:
I really really really do not want to **explicitly not solve** the problem here. This is also the first time I've been able to get *any eyes at all* actually looking at this code in the 15 months since I wrote it, so it would be really unfair to once again reject it out of hand without any opportunity to refactor.